File: leetcode/C++/best-time-to-buy-and-sell-stock-iii.cpp
class Solution {
public:
    int maxProfit(vector<int>& prices) {
        int sol=0, mn=prices[0];
        vector<int> mx(prices.size(),prices.back());
        if(prices.size()==1) return sol;
        for(int i=(int)prices.size()-2; i>=0; --i){
            mx[i] = max(mx[i+1],prices[i]);
            sol = max(sol,mx[i+1]-prices[i]);
        }
        for(int i=2, tr=prices[1]-prices[0]; i<(int)prices.size()-1; ++i){
            tr = max(tr,prices[i-1]-mn);
            mn = min(mn, prices[i-1]);
            sol = max(sol,tr + mx[i+1]-prices[i]);
        }
        return sol;
    }
};

File: leetcode/Python/best-time-to-buy-and-sell-stock-iii.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        sol, mx, mn = 0, [prices[-1]]*len(prices), prices[0]
        if len(prices) == 1:
            return sol
        # 1. update sol in case of only first transaction
        #    and update mx array (mx[i]: i~n max value) 
        for i in range(len(prices)-2, -1, -1):
            sol = max(sol, mx[i+1]-prices[i])
            mx[i] = max(mx[i+1], prices[i])
        # 2. update sol in case of second transaction
        # if second buy in idx i
        tr = prices[1] - prices[0]
        for i in range(2, len(prices)-1):
            tr = max(tr, prices[i-1] - mn)
            mn = min(mn, prices[i-1])
            sol = max(sol, tr + mx[i+1] - prices[i])
        return sol
            